syntax = "proto3";

import "common/Common.proto";


import "google/protobuf/descriptor.proto";

message ServerSideMessageOptions {  
	OMsgID id = 1;
}

extend google.protobuf.MessageOptions {
    ServerSideMessageOptions server_cmd = 1012;
}

package ServerMSG;


enum OMsgID
{
    OMSGID_NONE = 0;
    MsgID_BEGIN = 20000;
    
    //内部消息id
    MsgID_ServiceCmd          = 20001;
    MsgID_ServiceHttpRequest  = 20002;
    MsgID_ServiceHttpResponse = 20003;

    MsgID_MessageTransmit     = 20007; // 转发消息
    MsgID_ServiceReady        = 20010;
    MsgID_ServiceRegister     = 20011;

    MsgID_SocketConnect     = 20012; // sck->other 一个socket连接上来了
    MsgID_SocketClose       = 20013; // sck->other socket关闭
    MsgID_SocketChangeDest  = 20014; // other->sck 切换socket对应的服务器
    MsgID_SocketAuth        = 20015; // other->sck 设置socket的认证信息
    MsgID_SocketStartAccept = 20016; // world->sck 开始监听
    MsgID_SocketStopAccept  = 20017; // world->sck 关闭监听
    MsgID_SocketLogin       = 20018; // auth->world login

    

    MsgID_PlayerEnterZone       = 20021;
    MsgID_PlayerChangeZone      = 20022;
    MsgID_PlayerChangeZone_Data = 20023;
    MsgID_PlayerLogout          = 20024;

    MsgID_SystemVarChange = 20030;
    MsgID_SceneCreate     = 20031;
    MsgID_PhaseCreate     = 20032;
    MsgID_PhaseDestory    = 20033;

    MsgID_MonsterGen      = 20041;
    MsgID_MonsterDestory  = 20042;
    MsgID_ActiveGen       = 20043;
    MsgID_KillGen         = 20044;
    MsgID_MonsterGenOnce  = 20045;
    MsgID_MonsterGenMulti = 20046;

    MsgID_ActorCreate         = 20100;
    MsgID_ActorDestory        = 20101;
    MsgID_ActorMove           = 20102;
    MsgID_ActorCastSkill      = 20103;
    MsgID_ActorCastSkill_Fail = 20104;
    MsgID_ActorSetHide        = 20105;
    MsgID_SyncTaskPhase       = 20106;
    MsgID_AddTaskPhase        = 20107;
    MsgID_RemoveTaskPhase     = 20108;
    MsgID_ActorFlyTo          = 20109;
    MsgID_AOIChange           = 20110;


    MsgID_TeamCreate       = 20130;
    MsgID_TeamDestory      = 20131;
    MsgID_TeamQuit         = 20132;
    MsgID_TeamKickMember   = 20133;
    MsgID_TeamInviteMember = 20134;
    MsgID_TeamAcceptInvite = 20135;
    MsgID_TeamApplyMember  = 20136;
    MsgID_TeamAcceptApply  = 20137;
    MsgID_TeamNewLeader    = 20138;
    MsgID_TeamMemberInfoList   = 20139;
    MsgID_TeamAddMember    = 20140;
    MsgID_TeamDelMember    = 20141;

    MsgID_GuildCreate       = 20160;
    MsgID_GuildDestory      = 20161;
    MsgID_GuildQuit         = 20162;
    MsgID_GuildKickMember   = 20163;
    MsgID_GuildInviteMember = 20164;
    MsgID_GuildAcceptInvite = 20165;
    MsgID_GuildApplyMember  = 20166;
    MsgID_GuildAcceptApply  = 20167;
    MsgID_GuildNewLeader    = 20168;
    MsgID_GuildMemberInfoList   = 20169;
    MsgID_GuildAddMember    = 20170;
    MsgID_GuildDelMember    = 20171;

    MsgID_END = 30000;
};

message ServiceCmd
{
    option (server_cmd).id = MsgID_ServiceCmd;
    repeated string cmds = 1;
};

message ServiceHttpRequest
{
    option (server_cmd).id = MsgID_ServiceHttpRequest;
    uint64              uid   = 1;
    map<string, string> kvmap = 2;
};

message ServiceHttpResponse
{
    option (server_cmd).id = MsgID_ServiceHttpResponse;
    uint64 uid             = 1;
    int32  response_code   = 2;
    string response_reason = 3;
    bytes  response_txt    = 4;
};

message MessageTransmit
{
    option (server_cmd).id = MsgID_MessageTransmit;
    bytes raw_message_head = 3;
    bytes raw_message = 4;
};

message ServiceReady
{
    option (server_cmd).id = MsgID_ServiceReady;
    uint32 serverport = 1;
    bool   ready      = 2;
};

message ServiceRegister
{
    option (server_cmd).id = MsgID_ServiceRegister;
    uint32 serverport  = 1;
    uint64 update_time = 2;
}

message SocketConnect
{
    option (server_cmd).id = MsgID_SocketConnect;
    uint64 vs = 1;
}

message SocketClose
{
    option (server_cmd).id = MsgID_SocketClose;
    uint64 vs = 1;
}

message SocketChangeDest
{
    option (server_cmd).id = MsgID_SocketChangeDest;
    uint64 vs       = 1;
    uint32 destport = 2;
}

message SocketAuth
{
    option (server_cmd).id = MsgID_SocketAuth;
    uint64 vs = 1;
    bytes open_id = 2;
}

message SocketLogin
{
    option (server_cmd).id = MsgID_SocketLogin;
    uint64 vs = 1;
    bytes open_id = 2;
}

message SocketStartAccept 
{
    option (server_cmd).id = MsgID_SocketStartAccept;
}

message SocketStopAccept 
{
    option (server_cmd).id = MsgID_SocketStopAccept;
}

message PlayerEnterZone
{
    option (server_cmd).id = MsgID_PlayerEnterZone;
    uint64 idplayer = 1;
    uint64 socket   = 2;
    uint64 target_scene_idx = 4;
    float  posx     = 5;
    float  posy     = 6;
    float  face     = 7;
}

message PlayerChangeZone
{
    option (server_cmd).id = MsgID_PlayerChangeZone;
    uint64 idplayer = 1;
    uint64 socket   = 2;
    uint32 idzone   = 3;
    uint64 target_scene     = 4;

    float  posx     = 6;
    float  posy     = 7;
    float  range    = 8;
    float  face     = 9;
}

message PlayerChangeZone_Data
{
    option (server_cmd).id = MsgID_PlayerChangeZone_Data;
    uint64 idplayer = 1;
    uint64 socket   = 2;
    //其他的内存数据
}

message PlayerLogout
{
    option (server_cmd).id = MsgID_PlayerLogout;
    uint64 idplayer = 1;
    uint64 socket   = 2;
}

message SystemVarChange
{
    option (server_cmd).id = MsgID_SystemVarChange;
    uint32 keyidx = 1;
    uint32 idx    = 2;
    uint32 val    = 3;
    string str    = 4;
    enum SystemVarChangeType
    {
        SVCT_CREATE   = 0;
        SVCT_SET_DATA = 1;
        SVCT_ADD_DATA = 2;
        SVCT_SET_STR  = 3;
    };
    SystemVarChangeType type = 5;
}

message SceneCreate
{
    option (server_cmd).id = MsgID_SceneCreate;
    uint64 scene_id = 1;
};

message PhaseCreate
{
    option (server_cmd).id = MsgID_PhaseCreate;
    uint64 scene_id = 1;
    uint64 phase_id = 2;
};

message PhaseDestory
{
    option (server_cmd).id = MsgID_PhaseDestory;
    uint64 scene_id = 1;
    uint64 phase_id = 2;
};

message MonsterGen
{
    option (server_cmd).id = MsgID_MonsterGen;
    uint64 scene_id     = 1;
    uint64 phase_id     = 2;
    uint32 gen_id       = 3;
    uint32 monster_type = 4;
    float  posx         = 5;
    float  posy         = 6;
    uint32 camp_id      = 7;
    uint32 monster_id   = 8;
};

message MonsterGenMulti
{
    option (server_cmd).id = MsgID_MonsterGenMulti;
    uint64 scene_id     = 1;
    uint64 phase_id     = 2;
    uint32 gen_id       = 3;
    uint32 monster_type = 4;
    float  posx         = 5;
    float  posy         = 6;
    uint32 camp_id      = 7;
    uint32 monster_id   = 8;
    uint32 num          = 9;
    float  range        = 10;
};

message MonsterDestory
{
    option (server_cmd).id = MsgID_MonsterDestory;
    repeated uint64 monster_id = 1;
};

message MonsterGenOnce
{
    option (server_cmd).id = MsgID_MonsterGenOnce;
    uint64 scene_id = 1;
    uint32 gen_id   = 2;
    uint64 phase_id = 3;
};

message ActiveGen
{
    option (server_cmd).id = MsgID_ActiveGen;
    uint64 scene_id = 1;
    uint32 gen_id   = 2;
    bool   active   = 3;
};

message KillGen
{
    option (server_cmd).id = MsgID_KillGen;
    uint64 scene_id = 1;
    uint32 gen_id   = 2;
};

message AOIChange
{
    option (server_cmd).id = MsgID_AOIChange;
    uint64 actor_id  = 1;
    uint64 scene_idx  = 2;
    float posx = 3;
    float posy = 4;
    repeated uint64 actor_add = 5;
    repeated uint64 actor_del = 6;


};

message ActorCreate
{
    option (server_cmd).id = MsgID_ActorCreate;
    uint64 actor_id  = 1;
    uint64 scene_id  = 2;
    string name      = 4;
    uint32 actortype = 5; // ACTOR_TYPE
    uint32 baselook  = 6; //基本外观数据
    uint32 prof      = 7; //玩家代表职业, 怪物/宠物/NPC代表类型ID
    uint32 lev       = 8;
    uint32 campid    = 9;
    uint32 hp        = 10;
    uint32 hpmax     = 11;
    uint32 mp        = 12;
    uint32 mpmax     = 13;
    uint32 movespd   = 14;
    uint64 ownerid   = 15;
    
    float posx = 16;
    float posy = 17;
    float face = 18;

    uint32 monster_gen_id = 20;
    uint64 phase_id       = 21;
    bool   must_see       = 22;
    uint32 team_id        = 23;
    uint64 guild_id       = 24;
};

message ActorDestory
{
    option (server_cmd).id = MsgID_ActorDestory;
    uint64 actor_id = 1;
    bool dead = 2;
};

message ActorMove
{
    option (server_cmd).id = MsgID_ActorMove;
    uint64 actor_id = 1;
    float  x        = 2;
    float  y        = 3;
};

message ActorFlyTo
{
    option (server_cmd).id = MsgID_ActorFlyTo;
    uint64 actor_id = 1;
    float  x        = 2;
    float  y        = 3;
};

message ActorCastSkill
{
    option (server_cmd).id = MsgID_ActorCastSkill;
    uint64 actor_id  = 1;
    uint64 target_id = 2;
    uint32 skill_id  = 3;
};

message ActorCastSkill_Fail
{
    option (server_cmd).id = MsgID_ActorCastSkill_Fail;
    uint64 actor_id = 1;
};

message ActorSetHide
{
    option (server_cmd).id = MsgID_ActorSetHide;
    uint64 actor_id   = 1;
    int32  hide_count = 2;
};

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////

message TeamCreate
{
    option (server_cmd).id = MsgID_TeamCreate;
    uint64 team_id   = 1;
    uint64 leader_id = 2;
};

message TeamDestory
{
    option (server_cmd).id = MsgID_TeamDestory;
    uint64 team_id = 1;
};

message TeamQuit
{
    option (server_cmd).id = MsgID_TeamQuit;
    uint64 team_id     = 1;
    uint64 operator_id = 2;
};
message TeamKickMember
{
    option (server_cmd).id = MsgID_TeamKickMember;
    uint64 team_id     = 1;
    uint64 operator_id = 2;
    uint64 kick_id     = 3;
};

message TeamInviteMember
{
    option (server_cmd).id = MsgID_TeamInviteMember;
    uint64 team_id     = 1;
    uint64 operator_id = 2;
    uint64 invitee_id  = 3;
};
message TeamAcceptInvite
{
    option (server_cmd).id = MsgID_TeamAcceptInvite;
    uint64 team_id    = 1;
    uint64 inviter_id = 2;
    uint64 invitee_id = 3;
    bool   result     = 4;
};

message TeamApplyMember
{
    option (server_cmd).id = MsgID_TeamApplyMember;
    uint64 applicant_id  = 1;
    uint64 respondent_id = 2;
};
message TeamAcceptApply
{
    option (server_cmd).id = MsgID_TeamAcceptApply;
    uint64 applicant_id  = 1;
    uint64 respondent_id = 2;
    bool   result        = 3;
};

message TeamNewLeader
{
    option (server_cmd).id = MsgID_TeamNewLeader;
    uint64 team_id       = 1;
    uint64 operator_id   = 2;
    uint64 new_leader_id = 3;
};

message TeamAddMember
{
    option (server_cmd).id = MsgID_TeamAddMember;
    uint64 team_id   = 1;
    uint64 member_id = 2;
};

message TeamDelMember
{
    option (server_cmd).id = MsgID_TeamDelMember;
    uint64 team_id   = 1;
    uint64 member_id = 2;
};

message TeamMemberInfoList
{
    option (server_cmd).id = MsgID_TeamMemberInfoList;
    repeated TeamMemberInfo member_list = 1;
}
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


message GuildCreate
{
    option (server_cmd).id = MsgID_GuildCreate;
    uint64 guild_id   = 1;
    uint64 leader_id = 2;
    string guild_name = 3;
    string leader_name = 4;
    uint32 create_time = 5;
};

message GuildDestory
{
    option (server_cmd).id = MsgID_GuildDestory;
    uint64 guild_id = 1;
};

message GuildQuit
{
    option (server_cmd).id = MsgID_GuildQuit;
    uint64 guild_id     = 1;
    uint64 operator_id = 2;
};
message GuildKickMember
{
    option (server_cmd).id = MsgID_GuildKickMember;
    uint64 guild_id     = 1;
    uint64 operator_id = 2;
    uint64 kick_id     = 3;
};

message GuildInviteMember
{
    option (server_cmd).id = MsgID_GuildInviteMember;
    uint64 guild_id     = 1;
    uint64 operator_id = 2;
    uint64 invitee_id  = 3;
};
message GuildAcceptInvite
{
    option (server_cmd).id = MsgID_GuildAcceptInvite;
    uint64 guild_id    = 1;
    uint64 inviter_id = 2;
    uint64 invitee_id = 3;
    bool   result     = 4;
};

message GuildApplyMember
{
    option (server_cmd).id = MsgID_GuildApplyMember;
    uint64 applicant_id  = 1;
    uint64 respondent_id = 2;
};
message GuildAcceptApply
{
    option (server_cmd).id = MsgID_GuildAcceptApply;
    uint64 applicant_id  = 1;
    uint64 respondent_id = 2;
    bool   result        = 3;
};

message GuildNewLeader
{
    option (server_cmd).id = MsgID_GuildNewLeader;
    uint64 guild_id       = 1;
    uint64 operator_id   = 2;
    uint64 new_leader_id = 3;
    string new_leader_name = 4;
};

message GuildAddMember
{
    option (server_cmd).id = MsgID_GuildAddMember;
    uint64 guild_id   = 1;
    uint64 member_id = 2;
};

message GuildDelMember
{
    option (server_cmd).id = MsgID_GuildDelMember;
    uint64 guild_id   = 1;
    uint64 member_id = 2;
};

message GuildMemberInfoList
{
    option (server_cmd).id = MsgID_GuildMemberInfoList;
    repeated GuildMemberInfo member_list = 1;
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
message SyncTaskPhase
{
    option (server_cmd).id = MsgID_SyncTaskPhase;
    uint64          player_id     = 1;
    repeated uint64 task_phase_id = 2;
}

message AddTaskPhase
{
    option (server_cmd).id = MsgID_AddTaskPhase;
    uint64 player_id     = 1;
    uint64 task_phase_id = 2;
}

message RemoveTaskPhase
{
    option (server_cmd).id = MsgID_RemoveTaskPhase;
    uint64 player_id     = 1;
    uint64 task_phase_id = 2;
}